app/src/main/java/org/mozilla/focus/tabs/tabtray/TabsSessionModel.java

```diff
@@ -5,48 +5,185 @@
+import android.support.annotation.Nullable;
+import android.view.View;
+import android.webkit.GeolocationPermissions;
+import android.webkit.ValueCallback;
+import android.webkit.WebChromeClient;
 
 import org.mozilla.focus.BuildConfig;
 import org.mozilla.focus.tabs.Tab;
+import org.mozilla.focus.tabs.TabView;
+import org.mozilla.focus.tabs.TabsChromeListener;
 import org.mozilla.focus.tabs.TabsSession;
+import org.mozilla.focus.tabs.TabsViewListener;
 
+import java.util.ArrayList;
 import java.util.List;
 
 class TabsSessionModel implements TabTrayContract.Model {
     @NonNull
     private TabsSession tabsSession;
+    private OnTabModelChangedListener onTabModelChangedListener;
+
+    private List<Tab> tabs = new ArrayList<>();
 
     TabsSessionModel(@NonNull TabsSession tabsSession) {
         this.tabsSession = tabsSession;
     }
 
+    @Override
+    public void loadTabs(OnLoadCompleteListener listener) {
+        tabs.clear();
+        tabs.addAll(tabsSession.getTabs());
+
+        if (listener != null) {
+            listener.onLoadComplete();
+        }
+    }
+
     @Override
     public List<Tab> getTabs() {
-        return tabsSession.getTabs();
+        return tabs;
     }
 
     @Override
-    public int getCurrentTabPosition() {
-        return tabsSession.getTabs().indexOf(tabsSession.getFocusTab());
+    public Tab getFocusedTab() {
+        return tabsSession.getFocusTab();
     }
 
     @Override
-    public void switchTab(int tabIdx) {
-        final List<Tab> tabs = tabsSession.getTabs();
-        if (tabIdx < 0 || tabIdx >= tabs.size()) {
+    public void switchTab(int tabPosition) {
+        if (tabPosition >= 0 && tabPosition < tabs.size()) {
+            Tab target = tabs.get(tabPosition);
+            List<Tab> latestTabs = tabsSession.getTabs();
+            boolean exist = latestTabs.indexOf(target) != -1;
+            if (exist) {
+                tabsSession.switchToTab(target.getId());
+            }
+        } else {
             if (BuildConfig.DEBUG) {
-                throw new ArrayIndexOutOfBoundsException("index: " + tabIdx + ", size: " + tabs.size());
+                throw new ArrayIndexOutOfBoundsException("index: " + tabPosition + ", size: " + tabs.size());
             }
-            return;
         }
-
-        tabsSession.switchToTab(tabs.get(tabIdx).getId());
     }
 
     @Override
     public void removeTab(int tabPosition) {
-        final List<Tab> tabs = tabsSession.getTabs();
-        tabsSession.dropTab(tabs.get(tabPosition).getId());
+        if (tabPosition >= 0 && tabPosition < tabs.size()) {
+            tabsSession.dropTab(tabs.get(tabPosition).getId());
+        } else {
+            if (BuildConfig.DEBUG) {
+                throw new ArrayIndexOutOfBoundsException("index: " + tabPosition + ", size: " + tabs.size());
+            }
+        }
+    }
+
+    @Override
+    public void subscribe(final Observer observer) {
+        if (onTabModelChangedListener == null) {
```

Comment:
Is the listener can be ```subscribe``` once only? Because there's no set as null in ```unsubscribe```, the second pair of ```subscribe``` will not replace the ```observer``` reference in ```onTabModelChangeListener```

Reply:
Fixed, thanks!